library/footer.cpp

```diff
@@ -0,0 +1,324 @@
+			read_size = count;
+
+		err = __eblob_read_ll(fd, buffer, read_size, offset);
+		if (err)
+			break;
+
+		result = murmurhash(buffer, read_size, result);
+		count -= read_size;
+		offset += read_size;
+	}
+
+	return err;
+}
+
+/*
+ * chunked_footer_offset() - calculates chunked footer offset within record pointed by @wc.
+ *
+ * Returns footer offset within record.
+ */
+static inline uint64_t chunked_footer_offset(struct eblob_write_control *wc) {
+	/* size of one checksum */
+	static const size_t f_size = sizeof(uint64_t);
+	/* size of whole record without header and final checksum */
+	const uint64_t size = wc->total_size - sizeof(struct eblob_disk_control) - f_size;
+	/*
+	 * @size includes only size of chunks and size of checksums of these chunks,
+	 * therefore number of chunks can be calculated via division @size by
+	 * size of chunk plus size of checksum with rounding up.
+	 * It requires rounding up because last chunk can be less than EBLOB_CSUM_CHUNK_SIZE.
+	 */
+	const uint64_t chunks_count = ((size  - 1) / (EBLOB_CSUM_CHUNK_SIZE + f_size)) + 1;
+	/*
+	 * checksums are placed at the end of the entry,
+	 * so it's offset within entry is calculated as
+	 * total_size of the entry minus size of all checksums
+	 */
+	return wc->total_size - (chunks_count + 1) * f_size;
+}
+
+/*
+ * eblob_chunked_mmhash() - calculate chunked MurmurHash64A of record pointed by @key, @wc, @offset and @size.
+ * It calculates MurmurHash64A of only chunks that intersect record's part specified by @offset and @size.
+ *
+ * Results:
+ * Returns negative error value or zero on success
+ * @footers - calculated MurmurHash64A of chunks
+ * @footers_offset - offset of record's footer with corresponding checksums.
+ * @footers_offset can be used for reading and verifying on-disk checksums or for writing calculated checksums
+ */
+static int eblob_chunked_mmhash(struct eblob_backend *b, struct eblob_key *key, struct eblob_write_control *wc,
+                               const uint64_t offset, const uint64_t size,
+                               std::vector<uint64_t> &checksums, uint64_t &checksums_offset) {
+	int err = 0;
+	uint64_t first_chunk = offset / EBLOB_CSUM_CHUNK_SIZE;
+	uint64_t last_chunk = (offset + size - 1) / EBLOB_CSUM_CHUNK_SIZE + 1;
+	const uint64_t offset_max = wc->ctl_data_offset + wc->total_data_size + sizeof(struct eblob_disk_control);
+	const uint64_t data_offset = wc->ctl_data_offset + sizeof(struct eblob_disk_control);
+	checksums_offset = wc->ctl_data_offset + chunked_footer_offset(wc) + first_chunk * sizeof(uint64_t);
+
+	try {
+		checksums.resize(last_chunk - first_chunk, 0);
+	} catch (const std::exception &e) {
+		eblob_log(b->cfg.log, EBLOB_LOG_ERROR, "blob i%d: %s: %s: failed to allocate checksums: %s\n",
+		          wc->bctl->index, eblob_dump_id(key->id), __func__, e.what());
+		return -ENOMEM;
+	}
+
+	/* checksumming of the entry is disabled, so skip calculation of checksums */
+	if (wc->flags & BLOB_DISK_CTL_NOCSUM)
```

Comment:
`BLOB_DISK_CTL_NOCSUM` doesn't remove footer from entry. It just disables entry checksumming and verification, but it's footer should be correctly written with zero-filled values.